### PR TITLE
Metric multi

### LINKED ISSUE #1710: new (Multi)Metric that does linear combination of existing metrics
sort of like MultiEncoder, use existing metrics, and weight them, eg
`0.1*speedBasedMetric+0.9*altMAPE` 

Relevant: #1363 

edit: 
suggested approach: construct instances of desired (sub)metrics and manually assign them to the multimetric in description.py, this avoids complicated parsing of "definition-strings". Ie.:
`metric100=Metric(type='altMAPE', windowSize=100, ..)
metric1000=Metric(type='altMAPE', windowSize=1000,...)
multi=Metric(type='multi', weights=[0.2 0.8], metrics=[metric100 metric1000])`

edit: 
- [ ] still blocked by #1848

### PR DESCRIPTION
**Pull request by breznak**
new MetricMulti allows us to combine weighted values of many sub-metrics with different parameters. 
Also reuses shared MovingAverage code in metrics.
Implement `multi metric` + tests + cleanup

Fixes #1710

### CODE REVIEW

**Comment by breznak** [comment]
there's a bug in the test values, I can't see why..

**Comment by rhyolight** [comment]
@breznak Please don't assign PRs to milestones. I would only like issues to be assigned to milestones because it makes the roadmap cleaner.

**Comment by breznak** [comment]
Yay! 
@scottpurdy please review

**Comment by scottpurdy** [comment]
@chetan51 - can you take this one?

File: examples/opf/experiments/anomaly/temporal/hotgym/description.py [hunk 1]
```diff
@@ -403,7 +403,7 @@
                  params={'errorMetric': 'aae', 'window': 1000, 'steps': steps}))
   control['metrics'].append(
       MetricSpec(field=config['predictedField'], metric='multiStep',
-                 inferenceElement='multiStepBestPredictions',
+                 inferenceElement='anomalyScore',
```
Comment (chetan51) [inline examples/opf/experiments/anomaly/temporal/hotgym/description.py]: What is the motivation for this change?

File: nupic/frameworks/opf/metrics.py [hunk 2]
```diff
@@ -184,54 +186,14 @@ def getModule(metricSpec):
...
-
+  elif metricName == 'multi':
+    return MetricMulti(metricSpec)
   else:
     raise Exception("Unsupported metric type: %s" % metricName)
 
 ################################################################################
 #               Helper Methods and Classes                                    #
 ################################################################################
-class _MovingAverage(object):
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Nice cleanup!

File: nupic/frameworks/opf/metrics.py [hunk 3]
```diff
@@ -338,7 +300,7 @@ class AggregateMetric(MetricsIface):
   """
   ___metaclass__ = ABCMeta
 
-
+  #FIXME @abstractmethod - this should be marked abstract method and required to be implemented
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Are these separate issues filed to take care of these and the other FIXME?

File: nupic/frameworks/opf/metrics.py [hunk 4]
```diff
@@ -1459,7 +1424,63 @@ def addInstance(self, groundTruth, prediction, record = None):
...
 
+###################################################################################
+class MetricMulti(MetricsIface):
+  """Multi metric can combine multiple other (sub)metrics and 
+     weight them to provide combined score."""
 
+  def __init__(self, metricSpec):
+    """MetricMulti constructor using metricSpec is not allowed."""
+    raise ValueError("MetricMulti cannot be constructed from metricSpec string! "
+                     "Use MetricMulti(id,weights,metrics) constructor instead.")
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: The order of arguments here isn't consistent with the order in the actual constructor below.

File: nupic/frameworks/opf/metrics.py [hunk 5]
```diff
@@ -1459,7 +1424,63 @@ def addInstance(self, groundTruth, prediction, record = None):
...
+
+  def __init__(self, weights, metrics, id=None, window=None):
+    """MetricMulti 
+       @param id - unique name for this metric
+       @param weights - [list of floats] used as weights
+       @param (sub)metrics - [list of metrics] 
+       @param id - (opt) name of the metric
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Duplicate param

File: nupic/frameworks/opf/metrics.py [hunk 6]
```diff
@@ -1459,7 +1424,63 @@ def addInstance(self, groundTruth, prediction, record = None):
...
+       @param window - (opt) window size for moving average, or None
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Please indicate what will happen if `None` is passed

File: nupic/frameworks/opf/metrics.py [hunk 7]
```diff
@@ -1459,7 +1424,63 @@ def addInstance(self, groundTruth, prediction, record = None):
...
 
 
+  def addInstance(self, groundTruth, prediction, record = None):
+    err = 0.0
+    subResults = [m.addInstance(groundTruth, prediction, record) for m in self.metrics]
+    for i in xrange(len(self.weights)):
+      try:
+        err += subResults[i]*self.weights[i]
+      except(TypeError):
+        self.err = None
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Should this throw an exception?

File: nupic/frameworks/opf/metrics.py [hunk 8]
```diff
@@ -1459,7 +1424,63 @@ def addInstance(self, groundTruth, prediction, record = None):
...
+     weight them to provide combined score."""
 
+  def __init__(self, metricSpec):
+    """MetricMulti constructor using metricSpec is not allowed."""
+    raise ValueError("MetricMulti cannot be constructed from metricSpec string! "
+                     "Use MetricMulti(id,weights,metrics) constructor instead.")
+
+  def __init__(self, weights, metrics, id=None, window=None):
+    """MetricMulti 
+       @param id - unique name for this metric
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: What is this `id` used for?

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 9]
```diff
@@ -758,6 +758,34 @@ def testCustomErrorMetric(self):
...
+    multi = MetricMulti(weights=[0.2, 0.8], metrics=[metric10, metric1000], id='multi1')
+    multi.verbosity = 1
+    print multi 
+    # create reference metrics (must be diff from metrics above used in MultiMetric, as they keep history)
+    metric1000ref = getModule(ms1)
+    metric10ref = getModule(ms2)
+
+    
+    gt = [(i/4+1) for i in range(500, 1000)]
+    p = [i for i in range(500)]
```
Comment (chetan51) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: Can just be `p = range(500)`.

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py [hunk 10]
```diff
@@ -758,6 +758,34 @@ def testCustomErrorMetric(self):
...
+    # create multi metric
+    multi = MetricMulti(weights=[0.2, 0.8], metrics=[metric10, metric1000], id='multi1')
+    multi.verbosity = 1
+    print multi 
+    # create reference metrics (must be diff from metrics above used in MultiMetric, as they keep history)
+    metric1000ref = getModule(ms1)
+    metric10ref = getModule(ms2)
+
+    
+    gt = [(i/4+1) for i in range(500, 1000)]
```
Comment (chetan51) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: Isn't this equivalent to the simpler `gt = range(100, 200)`?

File: examples/opf/experiments/anomaly/temporal/hotgym/description.py (hunk 1 above)
Comment (breznak) [inline examples/opf/experiments/anomaly/temporal/hotgym/description.py]: some mistake, reverted. thanks

File: nupic/frameworks/opf/metrics.py (hunk 3 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: https://github.com/numenta/nupic/issues/1892

File: nupic/frameworks/opf/metrics.py (hunk 7 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: I think not, as it's the submetrics who can pass None, so I leave it upto the to handle that.

File: nupic/frameworks/opf/metrics.py (hunk 8 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: preparation for the next steps. in the `minimize=` field of `permutations.py` the MultiMetric cannot be described/constructed by the string, so I want to make it possible to identify the metric by `id`

File: tests/unit/nupic/frameworks/opf/opf_metrics_test.py (hunk 10 above)
Comment (breznak) [inline tests/unit/nupic/frameworks/opf/opf_metrics_test.py]: it's not. but actually the ranges do not matter much, I just copied them from other test.

**Comment by breznak** [comment]
@chetan51 thanks for review! fixed

File: nupic/frameworks/opf/metrics.py [hunk 11]
```diff
@@ -1459,7 +1424,62 @@ def addInstance(self, groundTruth, prediction, record = None):
...
 
+  def __init__(self, metricSpec):
+    """MetricMulti constructor using metricSpec is not allowed."""
+    raise ValueError("MetricMulti cannot be constructed from metricSpec string! "
+                     "Use MetricMulti(weights,metrics) constructor instead.")
+
+  def __init__(self, weights, metrics, id=None, window=None):
+    """MetricMulti 
+       @param weights - [list of floats] used as weights
+       @param (sub)metrics - [list of metrics]
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Replace with `@param metrics - [list of submetrics]` for clarity

File: nupic/frameworks/opf/metrics.py (hunk 7 above)
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: Sorry, I don't understand your reply, can you elaborate?

File: nupic/frameworks/opf/metrics.py [hunk 12]
```diff
@@ -1459,7 +1424,62 @@ def addInstance(self, groundTruth, prediction, record = None):
...
+  def __init__(self, metricSpec):
+    """MetricMulti constructor using metricSpec is not allowed."""
+    raise ValueError("MetricMulti cannot be constructed from metricSpec string! "
+                     "Use MetricMulti(weights,metrics) constructor instead.")
+
+  def __init__(self, weights, metrics, id=None, window=None):
+    """MetricMulti 
+       @param weights - [list of floats] used as weights
+       @param (sub)metrics - [list of metrics] 
+       @param id - (opt) name of the metric
```
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: @breznak explained what this `id` is for:

> preparation for the next steps. in the minimize= field of permutations.py the MultiMetric cannot be described/constructed by the string, so I want to make it possible to identify the metric by id

What about the other metrics? If `id` is optional, how do you plan to minimize over them?

File: nupic/frameworks/opf/metrics.py (hunk 7 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: I only return None when the existing metric would return None, so just keeping status quo..

File: nupic/frameworks/opf/metrics.py (hunk 12 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: other metrics can use the existing `minimize=<string>` form, multi metric cannot be constructed using this string, thus needs a special case - the id.

Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: I think we should consolidate them into a single approach for identifying metrics. Special cases can get messy. What would you propose as a single id mechanism?

**Comment by breznak** [comment]
@chetan51 thanks! how does it look now? :)

File: nupic/frameworks/opf/metrics.py (hunk 7 above)
Comment (chetan51) [inline nupic/frameworks/opf/metrics.py]: This try/catch is confusing and can introduce bugs that are hard to catch (if `subResults` contains a string instead of `None`, for example, this metric would silently return `None`). Please refactor.

File: nupic/frameworks/opf/metrics.py (hunk 12 above)
Comment (breznak) [inline nupic/frameworks/opf/metrics.py]: I already have a ticket for this, https://github.com/numenta/nupic/issues/1848 I think the `minimize=` should take an ID, rather then the current string description, as the latter leads to dumb bugs when one modifies his metric in description.py but forgets to do the same modification in permutation.py's minimize=. But that's not goal of this PR. I only include id in MetricMulti because it's new and can't be created from string anyway.

**Comment by chetan51** [comment]
I suggest that we either add `id` to all metrics, or none of them (for now). It's weird for just `MetricMulti` to have optional `id`, especially if it's not even being used anywhere.

**Comment by breznak** [comment]
@chetan51 agree. removed

**Comment by chetan51** [comment]
Great, looks good to me now. Thanks @breznak! Feel free to merge after tests pass.
